src/common/tensor_inspector.h

```diff
@@ -0,0 +1,763 @@
+        }
+        break;
+      case AbnormalChecker:
+        if (std::is_same<DType, float>::value || std::is_same<DType, double>::value ||
+            std::is_same<DType, mshadow::half::half_t>::value) {
+          return [] (DType x) {
+                return x == (DType)1.0 / (DType)0.0 || x == -(DType)1.0 / (DType)0.0 ||
+                    x != x;
+              };
+        } else {
+          LOG(WARNING) << "AbnormalChecker only applies to float types. " <<
+              "Lambda will always return false.";
+        }
+        break;
+      default:
+        return [] (DType x) {
+              return false;
+            };
+    }
+    return [] (DType x) {return false;};
+  }
+
+  /*!
+   * \brief calculate the coordinate of a value in the tensor, given its index
+   * \param idx the index of the value in the tensor
+   */
+  std::vector<int> index_to_coordinates(size_t idx) {
+    const int dimension = tb_.ndim();
+    std::vector<int> ret;
+    for (int i = dimension-1; i >= 0; --i) {
+      ret.push_back(idx % tb_.shape_[i]);
+      idx /= tb_.shape_[i];
+    }
+    std::reverse(ret.begin(), ret.end());
+    return ret;
+  }
+
+  /*!
+   * \brief check/validate the values within the tensor, find the coordinates
+   * where the value checker evaluates to true
+   * \tparam DType the data type
+   * \param ret a vector of coordinates which itself is a vector of int; calculated here
+   * \param checker the lambda function to check each value of within the tensor
+   * \param interactive wherether to allow the user to interactively check the coordinates
+   * \param tag the name given to this call
+   */
+  template<typename DType>
+  void check_value_helper(std::vector<std::vector<int>>* ret,
+      const std::function<bool(DType)>& checker, bool interactive, std::string tag) {
+#if MXNET_USE_CUDA
+    if (tb_.dev_mask() == gpu::kDevMask) {
+      return TensorInspector(test::CAccessAsCPU(ctx_, tb_, false)(), ctx_)
+          .check_value_helper<DType>(ret, checker, interactive, tag);
+    }
+#endif  // MXNET_USE_CUDA
+    size_t count = 0;
+    std::stringstream ss;
+    ss << "[";
+    bool first_pass = true;
+    for (size_t i = 0; i < tb_.shape_.Size(); ++i) {
+      if (checker(tb_.dptr<DType>()[i])) {
+        count += 1;
+        if (!first_pass) {
+          ss  << ", ";
+        }
+        first_pass = false;
+        std::vector<int> coords = index_to_coordinates(i);
+        ss << "(" << coords[0];
+        for (unsigned int i = 1; i < coords.size(); ++i) {
+          ss << ", " << coords[i];
+        }
+        ss << ")";
+        ret->push_back(coords);
+      }
+    }
+    ss << "]" << std::endl;
+    if (interactive) {
+      std::lock_guard<std::mutex> lock(InspectorManager::get()->mutex_);
+       InspectorManager::get()->check_value_tag_counter_[tag] += 1;
+      while (!InspectorManager::get()->check_value_skip_all_) {
+        std::cout << "----------Value Check----------" << std::endl;
```

Comment:
I got that covered already!

Reply:
Let's say we have a tensor of shape 20*1*5*5, then if you enter say `0,0` it will print a 5*5 tensor. If you enter `0,0,0,0` the if will print just a single value

![](https://user-images.githubusercontent.com/16669457/61092906-0f48e680-a3fd-11e9-8251-c4371cdd00ad.png)

Reply:
Good point. Will add this

Reply:
added!